File: web/lib/costConfig.ts
// Unified A-share transaction cost model.
// Reads from config/cost-model.json so TypeScript and Python share the same
// economic contract.  The file is resolved relative to the repo root, which
// is two levels up from web/lib/.
import fs from "node:fs";
import { repoConfigPath } from "./repoConfig";

export interface CostModel {
  buyCommissionBps: number;
  sellCommissionBps: number;
  stampDutyBps: number;
  baseSlippageBps: number;
  minimumCommissionYuan: number;
  impactModel: string;
  impactCoefficient: number;
  impactVolatilityLookback: number;
  maxImpactBps: number;
  missingTurnoverPenaltyBps: number;
}

export interface CostConfig {
  buyCommissionBps: number;
  sellCommissionBps: number;
  stampDutyBps: number;
  slippageBps: number;
  minimumCommissionYuan?: number;
  impactModel?: string;
  impactCoefficient?: number;
  impactVolatilityLookback?: number;
  maxImpactBps?: number;
  missingTurnoverPenaltyBps?: number;
}

let cached: CostModel | null = null;

export function loadCostModel(): CostModel {
  if (cached) return cached;
  const filePath = repoConfigPath("cost-model.json");
  const raw = JSON.parse(fs.readFileSync(filePath, "utf-8")) as Record<string, unknown>;
  const parsed: CostModel = {
    buyCommissionBps: Number(raw.buy_commission_bps),
    sellCommissionBps: Number(raw.sell_commission_bps),
    stampDutyBps: Number(raw.stamp_duty_bps),
    baseSlippageBps: Number(raw.base_slippage_bps),
    minimumCommissionYuan: Number(raw.minimum_commission_yuan),
    impactModel: String(raw.impact_model ?? "sqrt-volume"),
    impactCoefficient: Number(raw.impact_coefficient ?? 0),
    impactVolatilityLookback: Number(raw.impact_volatility_lookback ?? 20),
    maxImpactBps: Number(raw.max_impact_bps ?? 0),
    missingTurnoverPenaltyBps: Number(raw.missing_turnover_penalty_bps ?? 0),
  };
  for (const [key, value] of Object.entries(parsed)) {
    if (key === "impactModel") continue;
    if (!Number.isFinite(value) || Number(value) < 0) {
      throw new Error(`invalid cost model ${key}: ${String(value)}`);
    }
  }
  if (!Number.isInteger(parsed.impactVolatilityLookback) || parsed.impactVolatilityLookback < 2) {
    throw new Error("impactVolatilityLookback must be an integer >= 2");
  }
  if (parsed.impactModel !== "square-root-participation") {
    throw new Error(`unsupported impact model: ${parsed.impactModel}`);
  }
  cached = parsed;
  return cached;
}

/** Convert the unified model to the CostConfig shape used by BacktestConfig. */
export function toCostConfig(model: CostModel = loadCostModel()): CostConfig {
  return {
    buyCommissionBps: model.buyCommissionBps,
    sellCommissionBps: model.sellCommissionBps,
    stampDutyBps: model.stampDutyBps,
    slippageBps: model.baseSlippageBps,
    minimumCommissionYuan: model.minimumCommissionYuan,
    impactModel: model.impactModel,
    impactCoefficient: model.impactCoefficient,
    impactVolatilityLookback: model.impactVolatilityLookback,
    maxImpactBps: model.maxImpactBps,
    missingTurnoverPenaltyBps: model.missingTurnoverPenaltyBps,
  };
}

/** Round-trip cost in bps (buy + sell). */
export function roundTripBps(model: CostModel = loadCostModel()): number {
  return model.buyCommissionBps + model.baseSlippageBps +
    model.sellCommissionBps + model.stampDutyBps + model.baseSlippageBps;
}

File: research/ashare_research/cost_config.py
"""Unified A-share transaction cost model shared by TypeScript (web) and Python (research).

Reads from ``config/cost-model.json`` so both languages use the same economic
contract.  The file is resolved relative to the repo root.
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path


@dataclass(frozen=True)
class CostModel:
    buy_commission_bps: float
    sell_commission_bps: float
    stamp_duty_bps: float
    base_slippage_bps: float
    minimum_commission_yuan: float
    impact_model: str

    @property
    def buy_total_bps(self) -> float:
        return self.buy_commission_bps + self.base_slippage_bps

    @property
    def sell_total_bps(self) -> float:
        return self.sell_commission_bps + self.stamp_duty_bps + self.base_slippage_bps

    @property
    def round_trip_bps(self) -> float:
        return self.buy_total_bps + self.sell_total_bps

    @property
    def avg_side_bps(self) -> float:
        """Legacy ``fee_bps`` per-side equivalent (avg of buy and sell)."""
        return self.round_trip_bps / 2


def _repo_root() -> Path:
    # research/ashare_research/cost_config.py → research/ashare_research → research → repo root
    return Path(__file__).resolve().parents[2]


@lru_cache(maxsize=1)
def load_cost_model() -> CostModel:
    fp = _repo_root() / "config" / "cost-model.json"
    raw = json.loads(fp.read_text("utf-8"))
    return CostModel(
        buy_commission_bps=float(raw["buy_commission_bps"]),
        sell_commission_bps=float(raw["sell_commission_bps"]),
        stamp_duty_bps=float(raw["stamp_duty_bps"]),
        base_slippage_bps=float(raw["base_slippage_bps"]),
        minimum_commission_yuan=float(raw.get("minimum_commission_yuan", 5.0)),
        impact_model=str(raw.get("impact_model", "sqrt-volume")),
    )
